File: zendure/classes/read_zendure.php
<?php
/**
 * Zendure SolarFlow Status Reader - PHP Version
 * Fetches status from Zendure SolarFlow device and saves to JSON file
 * 
 * Usage:
 *   php read_zendure.php [ip_address]
 *   or use as a class in other scripts
 */

require_once __DIR__ . '/../includes/config_loader.php';

class SolarFlow2400 {
    private $url;
    private $ip;
    private $dataDir;
    private $dataFile;
    
    /**
     * Initialize the SolarFlow2400 reader
     * 
     * @param string $ipAddress IP address of the Zendure device
     */
    public function __construct($ipAddress) {
        $this->ip = $ipAddress;
        $this->url = "http://{$ipAddress}/properties/report";
        
        // Load config to get data directory
        $config = require __DIR__ . '/../includes/config_loader.php';
        // Resolve relative path (config uses '../data/' relative to zendure folder)
        $dataDirPath = $config['dataDir'];
        if (strpos($dataDirPath, '../') === 0) {
            // Relative path - resolve from zendure folder (go up one more level from classes)
            // __DIR__ is zendure/classes, dirname(__DIR__) is zendure, dirname(dirname(__DIR__)) is Energy
            $baseDir = dirname(dirname(__DIR__));
            $this->dataDir = $baseDir . '/' . trim(str_replace('../', '', $dataDirPath), '/');
        } else {
            // Absolute or relative to current directory
            $this->dataDir = $dataDirPath;
        }
        $this->dataFile = rtrim($this->dataDir, '/') . '/zendure_data.json';
    }
    
    /**
     * Write JSON data atomically to avoid concurrency issues.
     * Writes to a temporary file first, then atomically renames it.
     * 
     * @param array $data Data to write as JSON
     * @throws Exception If writing fails
     */
    private function writeJsonAtomic($data) {
        // Ensure data directory exists
        if (!is_dir($this->dataDir)) {
            if (!mkdir($this->dataDir, 0755, true)) {
                throw new Exception("Failed to create data directory: {$this->dataDir}");
            }
        }
        
        // Write to temporary file first
        $tempFile = $this->dataFile . '.tmp';
        $jsonContent = json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
        
        try {
            if (file_put_contents($tempFile, $jsonContent) === false) {
                throw new Exception("Failed to write temporary file: {$tempFile}");
            }
            
            // Atomically replace the target file with the temp file
            if (!rename($tempFile, $this->dataFile)) {
                throw new Exception("Failed to rename temporary file to target file");
            }
        } catch (Exception $e) {
            // Clean up temp file if something goes wrong
            if (file_exists($tempFile)) {
                @unlink($tempFile);
            }
            throw $e;
        }
    }
    
    /**
     * Fetch status from the Zendure device
     * 
     * @param bool $verbose If true, print detailed output
     * @return array|false Returns data array on success, false on failure
     */
    public function getStatus($verbose = true) {
        try {
            // Use cURL for HTTP request (more reliable than file_get_contents)
            $ch = curl_init($this->url);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_TIMEOUT, 5);
            curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);
            curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
            
            $response = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $curlError = curl_error($ch);
            curl_close($ch);
            
            if ($response === false || !empty($curlError)) {
                throw new Exception("cURL error: {$curlError}");
            }
            
            if ($httpCode !== 200) {
                throw new Exception("HTTP error: {$httpCode}");
            }
            
            $data = json_decode($response, true);
            if (json_last_error() !== JSON_ERROR_NONE) {
                throw new Exception("JSON decode error: " . json_last_error_msg());
            }
            
            // Extract properties and pack data
            $props = $data['properties'] ?? [];
            $packs = $data['packData'] ?? [];
            
            if ($verbose) {
                echo "API URL: {$this->url}\n\n";
                
                // Print all properties
                foreach ($props as $key => $value) {
                    echo "{$key}: {$value}\n";
                }
                
                echo "\n--- Zendure SolarFlow 2400 AC ({$this->ip}) ---\n";
                echo "System SoC:      " . ($props['electricLevel'] ?? 'N/A') . "%\n";
                echo "Solar Input:     " . ($props['solarInputPower'] ?? 0) . " W\n";
                echo "Home Output:     " . ($props['outputHomePower'] ?? 0) . " W\n";
                
                if (isset($props['BatVolt'])) {
                    echo "Battery Voltage: " . number_format($props['BatVolt'] / 100, 2) . " V\n";
                }
                
                if (isset($props['hyperTmp'])) {
                    echo "Unit Temp:       " . number_format($props['hyperTmp'] / 100, 1) . "°C\n";
                }
                
                echo "Grid Charge:     " . ($props['gridInputPower'] ?? 0) . " W\n";
                echo "Pack Input:      " . ($props['packInputPower'] ?? 0) . " W\n";
                
                echo "\n--- Battery Pack Details ---\n";
                foreach ($packs as $i => $pack) {
                    $packNum = $i + 1;
                    $sn = $pack['sn'] ?? 'N/A';
                    $level = $pack['socLevel'] ?? 'N/A';
                    $temp = isset($pack['maxTemp']) ? number_format($pack['maxTemp'] / 100, 1) : 'N/A';
                    $state = $pack['state'] ?? 'N/A';
                    echo "Pack {$packNum} [{$sn}]:\n";
                    echo "  Level: {$level}% | Temp: {$temp}°C | State: {$state}\n";
                }
                
                echo "--------------------------------------------\n";
            }
            
            // Prepare reading data with timestamp
            $readingData = [
                'timestamp' => date('c'), // ISO 8601 format
                'properties' => $props,
                'packData' => $packs
            ];
            
            // Save to JSON file atomically
            try {
                $this->writeJsonAtomic($readingData);
                if ($verbose) {
                    echo "✅ Reading saved to {$this->dataFile}\n";
                }
            } catch (Exception $e) {
                if ($verbose) {
                    echo "⚠️  Warning: Failed to save reading to file: {$e->getMessage()}\n";
                }
            }
            
            return $readingData;
            
        } catch (Exception $e) {
            if ($verbose) {
                echo "Error fetching data: {$e->getMessage()}\n";
            }
            return false;
        }
    }
}

// CLI usage
if (php_sapi_name() === 'cli') {
    // Load config to get default IP
    $config = require __DIR__ . '/../includes/config_loader.php';
    
    // Get IP address from command line argument or use config default
    $ipAddress = $argv[1] ?? $config['deviceIp'];
    
    if (empty($ipAddress)) {
        echo "Error: No IP address specified. Please set 'deviceIp' in config.php or provide as command line argument.\n";
        exit(1);
    }
    
    $solarflow = new SolarFlow2400($ipAddress);
    $result = $solarflow->getStatus(true);
    
    exit($result === false ? 1 : 0);
}
?>

File: read_zendure.py
import requests
import json
import os
from datetime import datetime

class SolarFlow2400:
    def __init__(self, ip_address):
        self.url = f"http://{ip_address}/properties/report"
        self.ip = ip_address
        self.data_dir = "data"
        self.data_file = os.path.join(self.data_dir, "zendure_data.json")
    
    def _write_json_atomic(self, data):
        """
        Write JSON data atomically to avoid concurrency issues.
        Writes to a temporary file first, then atomically renames it.
        """
        # Ensure data directory exists
        os.makedirs(self.data_dir, exist_ok=True)
        
        # Write to temporary file first
        temp_file = self.data_file + ".tmp"
        try:
            with open(temp_file, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            
            # Atomically replace the target file with the temp file
            os.replace(temp_file, self.data_file)
        except Exception as e:
            # Clean up temp file if something goes wrong
            if os.path.exists(temp_file):
                try:
                    os.remove(temp_file)
                except:
                    pass
            raise e

    def get_status(self):
        try:
            response = requests.get(self.url, timeout=5)
            response.raise_for_status()
            data = response.json()
            
            # The main system data is inside 'properties'
            props = data.get("properties", {})
            packs = data.get("packData", [])

            for key, value in props.items():
                print(f"{key}: {value}")

            print(f"\n--- Zendure SolarFlow 2400 AC ({self.ip}) ---")
            print(f"System SoC:      {props.get('electricLevel')}%")
            print(f"Solar Input:     {props.get('solarInputPower')} W")
            print(f"Home Output:     {props.get('outputHomePower')} W")
            print(f"Battery Voltage: {props.get('BatVolt') / 100:.2f} V")
            print(f"Unit Temp:       {props.get('hyperTmp') / 100:.1f}°C")

            print(f"Grid Charge:     {props.get('gridInputPower')} W")
            print(f"Pack Input:      {props.get('packInputPower')} W")
            
            print("\n--- Battery Pack Details ---")
            for i, pack in enumerate(packs):
                print(f"Pack {i+1} [{pack.get('sn')}]:")
                print(f"  Level: {pack.get('socLevel')}% | Temp: {pack.get('maxTemp')/100:.1f}°C | State: {pack.get('state')}")
            
            print("--------------------------------------------")
            
            # Save the reading to JSON file atomically
            reading_data = {
                "timestamp": datetime.now().isoformat(),
                "properties": props,
                "packData": packs
            }
            try:
                self._write_json_atomic(reading_data)
                print(f"✅ Reading saved to {self.data_file}")
            except Exception as e:
                print(f"⚠️  Warning: Failed to save reading to file: {e}")
            
            return reading_data
            
        except Exception as e:
            print(f"Error fetching data: {e}")
            return False

# Usage
my_solarflow = SolarFlow2400("192.168.2.93") # Use your specific IP
my_solarflow.get_status()